src/frontend/src/optimizer/plan_node/logical_join.rs

```diff
@@ -293,35 +290,53 @@ impl ColPrunable for LogicalJoin {
 
         let left_len = self.left.schema().fields.len();
 
-        let mut visitor = CollectInputRef::new(required_cols.clone());
+        let total_len = self.left().schema().len() + self.right().schema().len();
+        let mut resized_required_cols = FixedBitSet::with_capacity(total_len);
+
+        required_cols.ones().for_each(|i| {
+            if self.is_right_join() {
+                resized_required_cols.insert(left_len + i);
+            } else {
+                resized_required_cols.insert(i);
+            }
+        });
+
+        let mut visitor = CollectInputRef::new(resized_required_cols.clone());
         self.on.visit_expr(&mut visitor);
-        let left_right_required_cols = visitor.collect();
+        let left_right_input_cols = visitor.collect();
```

Comment:
Could you explain why did you change the variables' naming?

Reply:
I thought it was more readable while coding. But I think its fine now that I read again.